nixos/gnupg: drop references to gtk2; pinentry-*: build dependent variant with an override instead of choosing the output

SuperSandro2000 opened the pull request:
## Description of changes

This PR basically cleans up all #270266

see https://github.com/NixOS/nixpkgs/pull/270266#issuecomment-1871700542
Closes #276143
Closes #133156

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 committed:
nixos/gnupg: drop references to gtk2, fix ncurses typo

con-f-use commented:
Just wanted to mention again, that all the other pinentry flavors don't handle both tty and graphical prompts as well as gtk2. It's either or with them, so I don't think it should be dropped as a pinentry flavor.

SuperSandro2000 commented:
> all the other pinentry flavors don't handle both tty and graphical prompts as well as gtk2.

I am not sure how gtk2 is realted to tty/curses.



> It's either or with them, so I don't think it should be dropped as a pinentry flavor.

It is only dropped as a default entry. You can still configure it to use it.

SuperSandro2000 committed:
pinentry-*: build dependent variant with an override instead of choosing the output

We don't want to depend on the *enabled* flavors of pinentry but on all
*possible* ones.

SuperSandro2000 committed:
nixos/gnupg: choose the pinentry variant that has the selected flavor definitely enabled

roberth commented on pkgs/tools/security/pinentry/default.nix:
```diff
@@ -22,6 +22,7 @@ let
       lib.optionalString (flag != null)
         (lib.enableFeature (lib.elem f enabledFlavors) ("pinentry-" + flag));
 
+  # keep in sync with programs.gnupg.agent.pinentryFlavor
```
How about putting in this package an `assert lib.attrNames flavorInfo == import ./flavor-list.nix;`, so that you can import that file from NixOS?

ambroisie replied:
What about using a `passthru.flavors = lib.attrNames flavorInfo`, or is using a package's passthru attributes in NixOS options disallowed?

SuperSandro2000 replied:
I've implemented the second option as this added the least amount of code.

jtojnar replied:
Normally, any access to `pkgs` would be disallowed but `gnupg.nix` is opting out of sandboxing because the module was already using `passthru`:

https://github.com/NixOS/nixpkgs/blob/0b3fbec39d1426331e57c0e289d20a6911218c12/nixos/modules/programs/gnupg.nix#L238-L239

ElvishJerricco commented:
@SuperSandro2000 

> I am not sure how gtk2 is realted to tty/curses.

The `gtk2` pinentry flavor is the only one that will correctly prompt in the terminal if necessary, or graphically if available. e.g. If I SSH into my machine, I can still use `gpg` with the `gtk2` pinentry flavor, because it will put the prompt in the terminal. Meanwhile, it will still give a graphical prompt if I'm using it from the desktop environment. Other graphical pinentry flavors will *always* put up the graphical prompt on the desktop, even if you're logged in remotely with SSH.

lobre commented on nixos/modules/programs/gnupg.nix:
```diff
@@ -66,16 +64,17 @@ in
 
     agent.pinentryFlavor = mkOption {
-      type = types.nullOr (types.enum pkgs.pinentry.flavors);
+      # keep in sync with all possible favlors of pinentry
```
Typo: favlors -> flavors

ambroisie commented on nixos/modules/programs/gnupg.nix:
```diff
@@ -66,16 +64,17 @@ in
     agent.pinentryFlavor = mkOption {
-      type = types.nullOr (types.enum pkgs.pinentry.flavors);
+      # keep in sync with all possible favlors of pinentry
+      type = types.nullOr (types.enum [ "curses" "tty" "emacs" "gtk2" "gnome3" "qt" ]);
```
If we're changing this type, why not also allow packages ([similar to home-manager's `rbw` module](https://github.com/nix-community/home-manager/blob/4d8f90205c6c90be2e81d94d0e5eedf71c1ba34e/modules/programs/rbw.nix#L44)).

This would allow using other `pinentry` implementations, like [`pinentry-rofi`](https://github.com/plattfot/pinentry-rofi).

SuperSandro2000 replied:
Please open another PR with that new feature. I don't want to expand the scope of this PR even further.

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12101,11 +12101,11 @@ with pkgs;
 
   pinentry = libsForQt5.callPackage ../tools/security/pinentry { };
```
This will still bring in all the flavours (through `gnupg`) so the overrides below are pointless.

SuperSandro2000 replied:
That depends on GUI support which by default is only enabled on darwin.

jtojnar replied:
Oh, you are right. Looks like this changed two years ago (1ee8f771320b45651d21e4cdbf15141e1c143887). I wonder if it even works without agent then.

SuperSandro2000 replied:
Lets discover that in another PR to keep the scope manageable.

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12101,11 +12101,11 @@ with pkgs;
+  pinentry-emacs = pinentry.override { enabledFlavors = [ "emacs" ]; };
+  pinentry-gtk2 = pinentry.override { enabledFlavors = [ "gtk2" ]; };
+  pinentry-qt = pinentry.override { enabledFlavors = [ "qt" ]; };
+  pinentry-gnome3 = pinentry.override { enabledFlavors = [ "gnome3" ]; };
```
I vaguely recall that curses was supposed to be a fallback, if gui could not be started. But now the only reference I found is https://github.com/NixOS/nixpkgs/pull/71095#discussion_r334303343, which is not very informative.

roberth commented on nixos/modules/programs/gnupg.nix:
```diff
@@ -17,8 +17,6 @@ let
     || xserverCfg.desktopManager.plasma5.enable
     || xserverCfg.desktopManager.deepin.enable then
       "qt"
-    else if xserverCfg.desktopManager.xfce.enable then
-      "gtk2"
```
What's the reason to drop this rule?
I don't think xfce4 users want a terminal based pinentry, or do they?

roberth replied:
Taking a step back, it seems that this is more of a task for the individual desktop environment modules, to add the right pinentry flavor to `PATH` when initializing their session. That way it works right even if you have multiple installed desktop environments.
Also it solves the architectural flaw of having to enumerate all desktop environments here.
That doesn't have to be solved in this PR, but seems to make like a good next step.

con-f-use replied:
> What's the reason to drop this rule? I don't think xfce4 users want a terminal based pinentry, or do they?

They might, xfce users use ttys or go into a non-graphical environment from time to time.

SuperSandro2000 replied:
I think we can drop this change after 3cdf1705f5374d82d7fa208be92b86e6320ec3f8 which fixes the bug I avoided here.

SuperSandro2000 replied:
Just felt like doing this in this PR. I hope this is the last step to make this perfect and we don't increase the scope further.

SuperSandro2000 committed:
programs/gnupg: move agent.pinentryFlavor to the desktops

This avoids evaluating all the desktop environments at this place.

SuperSandro2000 commented:
I cannot test all the changes myself - I think they they are correct but I would appreciate it, if other people could help to verify that.

fpletz reviewed (approved):
Changes LGTM. Since I'm also using pinentry both graphically and with curses I'm gonna test and merge later today if nobody beats me to it.

fpletz reviewed (changes requested):
Do not merge for now since only setting one `enabledFlavor` probably does not provide a proper fallback.

SuperSandro2000 commented:
> Do not merge for now since only setting one `enabledFlavor` probably does not provide a proper fallback.

> I vaguely recall that curses was supposed to be a fallback, if gui could not be started. But now the only reference I found is [#71095 (comment)](https://github.com/NixOS/nixpkgs/pull/71095#discussion_r334303343), which is not very informative.

Based on this I am not sure that is even working.

fpletz commented:
> Based on this I am not sure that is even working.

I'm using this functionality daily and others have also noted they are using this in other PRs regarding the breakage of pinenentry-gtk2.

rnhmjoj reviewed (approved):
Looks good to me. I like that DEs set the flavour they need, instead of having gnupg handling it for each and every one.

SuperSandro2000 commented:
> I'm using this functionality daily and others have also noted they are using this in other PRs regarding the breakage of pinentry-gtk2.

yeah, gtk2 but we don't really know that for the other flavors and I kinda remember that jan some time said, that this might not work.
